### PR TITLE
Total Refactor: SDK v0 & v1 APIs, Unified Webhooks & Struct Optimization

### PR DESCRIPTION
**Pull request by amatsagu**
> [!WARNING]
> This PR completely revamps the SDK to support both v0 and the latest v1 Top.gg API endpoints. I've improvized with some elements due to unclear source od truth (hard to navigate, jumbled docs page).

Brings the entire SDK up to date with the newest Top.gg v1 endpoints and fully optimizes memory footprints.

For example, our new Webhook Handler dynamically sniffs HTTP headers to parse and auto-normalize legacy `v0` payloads directly into the strict `v1` `VoteCreatePayload` structure, ensuring that your logic callbacks only ever need to be written once:
```go
webhookHandler := client.NewWebhookHandler(dbl.WebhookOptions{
	Secret: "YOUR_WEBHOOK_SECRET",
	OnVote: func(vote dbl.VoteCreatePayload) {
		fmt.Printf("Received vote from user %s with weight %d\n", vote.User.ID, vote.Weight)
	},
})
```

> [!NOTE]
> All structs have been run through `fieldalignment` to drastically reduce byte padding and optimize heap allocations. Additionally, raw string mappings for Platforms, Project Types, and Locales have been replaced with Enums.

<details>
<summary><b>Supported API Endpoints</b></summary>
<br>
<ul>
<li><input type="checkbox" checked disabled> <code>[GET] /v1/projects/{project_id}</code> - GetProject</li>
<li><input type="checkbox" checked disabled> <code>[GET] /v1/projects/@me</code> - GetMyProject</li>
<li><input type="checkbox" checked disabled> <code>[PATCH] /v1/projects/@me</code> - EditMyProject</li>
<li><input type="checkbox" checked disabled> <code>[PUT] /v1/projects/@me/commands</code> - PostApplicationCommands</li>
<li><input type="checkbox" checked disabled> <code>[POST] /v1/projects/@me/announcements</code> - PostAnnouncement</li>
<li><input type="checkbox" checked disabled> <code>[PATCH] /v1/projects/@me/metrics</code> - PostMyMetrics</li>
<li><input type="checkbox" checked disabled> <code>[POST] /v1/projects/@me/metrics/batch</code> - PostMyMetricsInBatch</li>
<li><input type="checkbox" checked disabled> <code>[GET] /v1/projects/@me/votes/{user_id}</code> - GetVote</li>
<li><input type="checkbox" checked disabled> <code>[GET] /v1/projects/@me/votes</code> - GetVotes</li>
<li><input type="checkbox" checked disabled> <code>[GET] /v0/bots/{id}</code> - GetBot</li>
<li><input type="checkbox" checked disabled> <code>[GET] /v0/bots</code> - GetBots</li>
<li><input type="checkbox" checked disabled> <code>[GET] /v0/bots/{id}/stats</code> - GetBotStats</li>
<li><input type="checkbox" checked disabled> <code>[POST] /v0/bots/{id}/stats</code> - PostBotStats</li>
<li><input type="checkbox" checked disabled> <code>[GET] /v0/bots/{id}/check</code> - CheckIfVoted</li>
<li><input type="checkbox" checked disabled> <code>[GET] /v0/users/{id}</code> - GetUser</li>
<li><input type="checkbox" checked disabled> <code>Unified Webhook Handler (v0 + v1 support)</code></li>
</ul>
</details>

<details>
<summary><b>Supported Structs & Types</b></summary>
<br>
<ul>
<li><input type="checkbox" checked disabled> <b>Bot Data:</b> <code>Bot</code>, <code>BotStats</code></li>
<li><input type="checkbox" checked disabled> <b>User Data:</b> <code>User</code>, <code>Social</code></li>
<li><input type="checkbox" checked disabled> <b>Project Data:</b> <code>Project</code>, <code>ProjectPayload</code>, <code>Announcement</code>, <code>MetricsPayload</code></li>
<li><input type="checkbox" checked disabled> <b>Vote Data:</b> <code>Vote</code>, <code>PartialVote</code>, <code>PaginatedVotes</code></li>
<li><input type="checkbox" checked disabled> <b>Webhook Payloads:</b> <code>WebhookPayload</code>, <code>VoteCreatePayload</code>, <code>WebhookTestPayload</code>, <code>IntegrationCreatePayload</code>, <code>IntegrationDeletePayload</code>, <code>v0VotePayload</code></li>
<li><input type="checkbox" checked disabled> <b>Webhook Sub-Entities:</b> <code>PartialProject</code>, <code>WebhookUser</code>, <code>WebhookOptions</code></li>
<li><input type="checkbox" checked disabled> <b>Type-Safe Enums:</b> <code>Platform</code>, <code>ProjectType</code>, <code>Locale</code>, <code>Scope</code></li>
</ul>
</details>

### CODE REVIEW

File: bot.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
+package dbl
```
Comment (velddev) [inline bot.go]: if this is going to be a complete re-do can we make a new package and call it topgg? I don't want old branding to still stick around 😄

File: bot.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+	Owners           []Snowflake `json:"owners,omitzero"`
+	ServerCount      int         `json:"server_count,omitempty"`
+	ShardCount       int         `json:"shard_count,omitempty"`
+	ID               Snowflake   `json:"id"`
+	Points           int         `json:"points"`
+	MonthlyPoints    int         `json:"monthlyPoints"`
+	CertifiedBot     bool        `json:"certifiedBot"`
+}
+
+// https://docs.top.gg/api/v0/bots#get-/bots/bot_id/stats
```
Comment (velddev) [inline bot.go]: preferably drop v0 support. It's end-of-life

File: bot.go (hunk 1 above)
Comment (amatsagu) [reply bot.go]: Changed dbl -> topgg

File: bot.go (hunk 2 above)
Comment (amatsagu) [reply bot.go]: Following request - I've ripped all code related to v0 API and slightly updated README examples.

File: go.mod [hunk 3]
```diff
@@ -1,8 +1,62 @@
-module github.com/top-gg/go-dbl
+module github.com/top-gg-community/go-sdk
 
-go 1.15
+tool github.com/evilmartians/lefthook/v2
```
Comment (DevYukine) [inline go.mod]: This pulls lefthook plus ~55 indirect modules into the SDK's go.mod and go.sum. Users importing our library don't build them, but they end up in their module graph, can bump shared deps via MVS, and trip security scanners. This library should have zero deps. Drop the tool directive and document go install for lefthook, or move it to a separate tools module imo

File: go.mod [hunk 4]
```diff
@@ -1,8 +1,62 @@
...
+
+go 1.26.4
```
Comment (DevYukine) [inline go.mod]: Don't pin a patch version here, it forces every user importing our library onto 1.26.4+. Declare the minimum you actually need, e.g. go 1.24.

File: webhook.go [hunk 5]
```diff
@@ -1,110 +1,207 @@
...
+		if w.onVote != nil {
+			var vote VoteCreatePayload
+			if err := json.Unmarshal(payload.Data, &vote); err == nil {
+				w.onVote(vote)
+			}
+		}
+	case ScopeIntegrationCreate:
+		var integration IntegrationCreatePayload
+		if err := json.Unmarshal(payload.Data, &integration); err == nil {
+			w.secret = integration.Secret // Auto-update secret
```
Comment (DevYukine) [inline webhook.go]: Data race. ServeHTTP runs concurrently and validateV1 reads w.secret while this writes it. Guard it with a RWMutex or atomic.Value. Also worth a doc note that the updated secret only lives in memory and is lost on restart.

File: rate-limit.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+		bucket.Remaining = bucket.Limit
+		if bucketKey == "global" {
+			bucket.ResetAt = time.Now().Add(time.Second)
+		} else {
+			bucket.ResetAt = time.Now().Add(time.Minute)
+		}
+	}
+
+	bucket.Remaining--
+	return func() {
```
Comment (DevYukine) [inline rate-limit.go]: This unlock only runs after RoundTrip returns, so the bucket mutex is held for the whole HTTP request. Everything is in the global bucket now, so the client does one request at a time. Line 95 also sleeps while holding the lock, and the global wait (up to 60s after a 429) ignores req.Context() while the client timeout is 10s, so long waits always end in a timeout. Decrement Remaining, release the lock before the request, and make the waits context aware.

File: rate-limit.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+func NewRateLimiter(opt RateLimiterOptions) *RateLimiter {
+	rl := &RateLimiter{
+		buckets:     make(map[string]*Bucket),
+		traceLogger: opt.TraceLogger,
+	}
+
+	rl.buckets["global"] = &Bucket{Limit: 100, Remaining: 100}
+	rl.buckets["/bots"] = &Bucket{Limit: 60, Remaining: 60}
```
Comment (DevYukine) [inline rate-limit.go]: Dead code after the v0 removal: these /bots and /users buckets and the /api/v0 stripping in RoundTrip never match anything. Same for Webhook.client, ErrRequireAuthentication, and WebhookPayload.Trace, which nothing populates.

File: client.go [hunk 8]
```diff
@@ -1,79 +1,232 @@
...
+	return c.trippedUntil.Load() > time.Now().UnixNano()
 }
 
-// NewClient returns a new *Client after applying the options provided.
-func NewClient(token string, options ...OptionFunc) (*Client, error) {
-	client := &Client{
-		limiter:    rate.NewLimiter(1, 60),
-		httpClient: &http.Client{Timeout: defaultTimeout},
-		token:      token,
+func (c *Client) request(method, route string, jsonPayload any) ([]byte, error) {
```
Comment (DevYukine) [inline client.go]: Since this is already a breaking rewrite, i would include context.Context in the public API now. Callers currently can't cancel or set deadlines, which matters with the sleeps in the rate limiter. Probably a good idea to add now as otherwise we need a major release because of breaking API

File: webhook.go [hunk 9]
```diff
@@ -1,110 +1,207 @@
...
 
-	body, err := ioutil.ReadAll(r.Body)
-
-	if err != nil {
-		return
+	switch payload.Type {
+	case ScopeVoteCreate:
+		if w.onVote != nil {
+			var vote VoteCreatePayload
+			if err := json.Unmarshal(payload.Data, &vote); err == nil {
```
Comment (DevYukine) [inline webhook.go]: Unmarshal errors here (and in the cases below) are ignored, even when i agree that we should return a 2XX on parsing errors as its likely a fault on our side i think we should at least log it or inform the user using this library about it? kinda feels weird to silently drop it

File: project.go [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+// https://docs.top.gg/api/v1/projects#post-/projects/@me/metrics/batch
+func (c *Client) PostMyMetricsInBatch(payload []MetricsPayload) error {
+	body := map[string]any{"data": payload}
+	_, err := c.request(http.MethodPost, "/v1/projects/@me/metrics/batch", body)
+	return err
+}
+
+// https://docs.top.gg/api/v1/votes#get-/projects/@me/votes/user_id
+func (c *Client) GetVote(userID Snowflake, source string) (*PartialVote, error) {
+	b, err := c.request(http.MethodGet, fmt.Sprintf("/v1/projects/@me/votes/%s?source=%s", userID, source), nil)
```
Comment (DevYukine) [inline project.go]: source is not escaped and ?source= is sent even when empty. Use url.Values like GetVotes does. Consider typing the param as Platform, the README shows you passing topgg.PlatformDiscord and that doesn't compile against string

File: project.go [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+	err = json.Unmarshal(b, &vote)
+	return &vote, err
+}
+
+// https://docs.top.gg/api/v1/votes#get-/projects/@me/votes
+func (c *Client) GetVotes(cursor string, startDate *time.Time) (*PaginatedVotes, error) {
+	q := url.Values{}
+	if startDate != nil {
+		q.Set("startDate", startDate.Format(time.RFC3339))
+	} else if cursor != "" {
```
Comment (DevYukine) [inline project.go]: If both cursor and startDate are set, the cursor is silently dropped. Return an Error, document the precedence, or send both.

File: README.md [hunk 12]
```diff
@@ -1,134 +1,174 @@
...
-)
+#### Discord ID
 
-func main() {
-	dblClient, err := dbl.NewClient("token")
-	if err != nil {
-		log.Fatalf("Error creating new Discord Bot List client: %s", err)
-	}
+```go
+vote, err := client.GetVote(topgg.Snowflake("661200758510977084"), topgg.PlatformDiscord)
```
Comment (DevYukine) [inline README.md]: This example doesn't compile i think?. Snowflake("...") is an invalid string to uint64 conversion (use StringToSnowflake) and PlatformDiscord doesn't match the string source param as i also said in another comment on the relevant code.

File: const.go [hunk 13]
```diff
@@ -1,5 +1,6 @@
-package dbl
+package topgg
 
 const (
-	BaseURL = "https://top.gg/api/"
+	BaseURL       = "https://top.gg/api"
+	DISCORD_EPOCH = 1420070400000
```
Comment (DevYukine) [inline const.go]: This is out of style for go const, i would stay consistent with DiscordEpoch here

**Review by DevYukine — CHANGES_REQUESTED** [review]
Generally a good direction but i would likely fix those before its ready to be merged

Oh and also it would be a good idea to set a user-agent for this library, should generally be good practise to set those

**Comment by amatsagu** [comment]
Quick warning - I did not manually verify if each endpoint works, just conceptually it should (assuming top.gg docs page is correct).

I left message on discord - https://discord.com/channels/264445053596991498/844289669118427226/1525799175529431112, regarding earlier code comments.

There are some things unclear so I'm left guessing. For example - Go packages should always have clear versioning model, easiest and most common one are [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/). There seems to be no releases done before + we have now v1 API so how about releasing later v1.0.0?

Other things I'm left guessing is how you want errors to be structured + traces. I did my own light mix, so code returns mostly pure error values and also produces some traces but those are on lightweight side - I personally dislike spamming dozens of traces per each action...

For time being, I moved lib Go versioning back to v1.9 and wrote code in style I remmeber seeing around that time. I'll wait for reply, feel free to criticize things, this is still rushed code after all, also I did ask for getting dev/testing API access secrets but received none so it did somewhat limit my tests.

File: project.go [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+}
+
+// https://docs.top.gg/api/v1/projects#patch-/projects/@me/metrics
+func (c *Client) PostMyMetrics(ctx context.Context, payload MetricsPayload) error {
+	_, err := c.request(ctx, http.MethodPatch, "/v1/projects/@me/metrics", payload)
+	return err
+}
+
+// https://docs.top.gg/api/v1/projects#post-/projects/@me/metrics/batch
+func (c *Client) PostMyMetricsInBatch(ctx context.Context, payload []MetricsPayload) error {
```
Comment (DevYukine) [inline project.go]: The batch endpoint expects each item to be `{timestamp?, metrics: {...}}`, but this sends raw metric objects.

File: rate-limit.go [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+			bodyBytes, err := io.ReadAll(resp.Body)
+			if cErr := resp.Body.Close(); cErr != nil {
+				lastResp = resp
+				lastErr = fmt.Errorf("%w: failed to close rate limit response body: %w", ErrRequestFailed, cErr)
+				time.Sleep(time.Millisecond * time.Duration(250*int64(i+1)))
+				continue
+			}
+
+			if err == nil {
+				var rateErr rateLimitError
```
Comment (DevYukine) [inline rate-limit.go]: `Retry-After` is a response header, not part of the JSON body.

File: rate-limit.go [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+					}
+
+					return nil, errRet
+				}
+			}
+
+			bodyBytes, err := io.ReadAll(resp.Body)
+			if cErr := resp.Body.Close(); cErr != nil {
+				lastResp = resp
+				lastErr = fmt.Errorf("%w: failed to close rate limit response body: %w", ErrRequestFailed, cErr)
```
Comment (DevYukine) [inline rate-limit.go]: This fails `go test` and `go vet` on Go 1.19 because `%w` is used twice. The same issue exists on lines 229 and 235.

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+- Go 1.19 or later (downloadable from [the official Go website](https://go.dev/dl/))
+- A suitable version of `golangci-lint` installed on your device (instructions on [their website](https://golangci-lint.run/docs/welcome/install/local/))
+- The repository [forked](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/fork-a-repo) and [cloned](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository) locally on your device
+
+## Lefthook Setup
+This project uses [Lefthook](https://lefthook.dev/) for development to run pre-commit hooks for linting and formatting. \
+It is **not** included as a tool dependency in `go.mod`, so it needs a separate command to initialize the various hooks:
+
+```bash
+go tool github.com/evilmartians/lefthook/v2 install
```
Comment (DevYukine) [inline CONTRIBUTING.md]: `go tool ...` no longer works because Lefthook is not declared as a tool. Use `go install ...@version`, then `lefthook install`.

File: project.go [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+}
+
+// https://docs.top.gg/api/v1/projects#discord-server
+// https://docs.top.gg/api/v1/projects#roblox-game
+type MetricsPayload struct {
+	ServerCount int `json:"server_count,omitempty"`
+	ShardCount  int `json:"shard_count,omitempty"`
+	MemberCount int `json:"member_count,omitempty"`
+	OnlineCount int `json:"online_count,omitempty"`
+	PlayerCount int `json:"player_count,omitempty"`
```
Comment (DevYukine) [inline project.go]: I'm not sure if this should be omitempty, i think player count being 0 instead of being absent is valid as roblox games can have 0 palyers? consider if you want to extend that to those other values too or not even tho server, shard and member count should really not be 0 when its listed on top.gg

File: rate-limit.go [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+
+			if !retrying {
+				retrying = true
+				if t.retryCounter.Add(1) > t.retryThreshold {
+					t.trippedUntil.Store(time.Now().Add(5 * time.Second).UnixNano())
+					return nil, fmt.Errorf("%w: global retry threshold (%d) exceeded", ErrLocalRatelimit, t.retryThreshold)
+				}
+			}
+
+			time.Sleep(time.Millisecond * time.Duration(250*int64(i+1)))
```
Comment (DevYukine) [inline rate-limit.go]: nit: maybe we can use a `select` here too so the request context can cancel the wait.

File: .github/workflows/go.yml [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+
+  build:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v6
+
+    - name: Set up Go
+      uses: actions/setup-go@v6
+      with:
+        go-version: '1.26.4'
```
Comment (DevYukine) [inline .github/workflows/go.yml]: nit: can we test against Go 1.19 here since that is the minimum version in `go.mod`?

**Review by DevYukine — CHANGES_REQUESTED** [review]

File: rate-limit.go (hunk 16 above)
Comment (amatsagu) [reply rate-limit.go]: Solved, weird random thing between versions...

File: .github/workflows/go.yml (hunk 20 above)
Comment (amatsagu) [reply .github/workflows/go.yml]: Changed - but this also required pinning specific versions of used tools because otherwise, @latest would panic it uses too old version of Go.

File: project.go (hunk 18 above)
Comment (amatsagu) [reply project.go]: Unless it is confirmed to cause issue - I would leave it as is. Those are integer values instead pointers so Go will automatically use 0 as default value. Moment player count is not specified, it will just become 0 on it's own. I have no details how it's handled on other side (Top.gg API) but I assume they treat lack of value as 0 too (as it's logical, it would be really weird if top.gg system allows to have a game with undefined / unknown player count).

File: CONTRIBUTING.md (hunk 17 above)
Comment (amatsagu) [reply CONTRIBUTING.md]: Fixed.

File: rate-limit.go (hunk 15 above)
Comment (amatsagu) [reply rate-limit.go]: Fixed.

File: project.go (hunk 14 above)
Comment (amatsagu) [reply project.go]: I believe it's now fixed but would like to see cleaner version of top.gg docs in future...

File: rate-limit.go (hunk 19 above)
Comment (amatsagu) [reply rate-limit.go]: I've tried making all places related to sleeping, giving time offset to be now context aware.

File: project.go (hunk 18 above)
Comment (DevYukine) [reply project.go]: `omitempty` removes zero-valued ints from the JSON. `MetricsPayload{PlayerCount: 0}` becomes `{}`, while the API requires `player_count` and allows 0. We need to distinguish unset from zero here. Please remove it from all the counts.

Comment (amatsagu) [reply project.go]: This feels incorrect but if it has to be this way... - fixed.

File: project.go [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+// https://docs.top.gg/api/v1/projects#patch-/projects/@me/metrics
+func (c *Client) PostMyMetrics(ctx context.Context, payload MetricsPayload) error {
+	_, err := c.request(ctx, http.MethodPatch, "/v1/projects/@me/metrics", payload)
+	return err
+}
+
+// https://docs.top.gg/api/v1/projects#post-/projects/@me/metrics/batch
+func (c *Client) PostMyMetricsInBatch(ctx context.Context, payload []BatchMetricsPayload) error {
```
Comment (velddev) [inline project.go]: why are some of these `XMyY` and some are just `XY`? would be nice to make it consistent

**Review by velddev — APPROVED** [review]

File: const.go [hunk 22]
```diff
@@ -1,5 +1,7 @@
-package dbl
+package topgg
 
 const (
-	BaseURL = "https://top.gg/api/"
+	BaseURL      = "https://top.gg/api"
+	UserAgent    = "Top.gg SDK https://github.com/top-gg-community/go-sdk"
```
Comment (velddev) [inline const.go]: This would follow the user-agent spec more closely (https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/User-Agent)
```suggestion
UserAgent = fmt.Sprintf(
    "topgg-go-sdk/%s (+https://github.com/top-gg-community/go-sdk)",
    Version,
)
```

Comment (amatsagu) [reply const.go]: Fixed.

File: project.go (hunk 21 above)
Comment (amatsagu) [reply project.go]: Fixed.

**Review by velddev — APPROVED** [review]
